File: apps/catalog/uom.py
"""Catálogo de unidades de medida (UOM) para las líneas de BOM.

Códigos de 2 caracteres alfanuméricos. Se usan tanto en el BOM de la EMPRESA
(BOMComponent) como en el del PROVEEDOR (SolicitationBOMLine), en la captura
manual (desplegable) y en la carga por layout. El catálogo del front
(frontend/src/lib/uom.ts) debe reflejar esta misma lista."""

UOM_CHOICES = [
    ("PZ", "Pieza"),
    ("UN", "Unidad"),
    ("KG", "Kilogramo"),
    ("GR", "Gramo"),
    ("MG", "Miligramo"),
    ("TN", "Tonelada"),
    ("LB", "Libra"),
    ("OZ", "Onza"),
    ("MT", "Metro"),
    ("CM", "Centímetro"),
    ("MM", "Milímetro"),
    ("M2", "Metro cuadrado"),
    ("M3", "Metro cúbico"),
    ("FT", "Pie"),
    ("IN", "Pulgada"),
    ("LT", "Litro"),
    ("ML", "Mililitro"),
    ("GL", "Galón"),
    ("PR", "Par"),
    ("JG", "Juego"),
    ("KT", "Kit"),
    ("CJ", "Caja"),
    ("RL", "Rollo"),
    ("HJ", "Hoja"),
]

UOM_CODES = {code for code, _ in UOM_CHOICES}


def clean_uom(value):
    """Normaliza un valor a un código UOM válido (2 caracteres en mayúsculas).
    Devuelve "" si está vacío o no pertenece al catálogo."""
    if not value:
        return ""
    code = "".join(ch for ch in str(value).strip().upper() if ch.isalnum())[:2]
    return code if code in UOM_CODES else ""

File: frontend/src/lib/uom.ts
// Catálogo de unidades de medida (UOM) para las líneas de BOM.
// Códigos de 2 caracteres alfanuméricos. Refleja apps/catalog/uom.py.
export const UOM_OPTIONS: { code: string; label: string }[] = [
  { code: "PZ", label: "Pieza" },
  { code: "UN", label: "Unidad" },
  { code: "KG", label: "Kilogramo" },
  { code: "GR", label: "Gramo" },
  { code: "MG", label: "Miligramo" },
  { code: "TN", label: "Tonelada" },
  { code: "LB", label: "Libra" },
  { code: "OZ", label: "Onza" },
  { code: "MT", label: "Metro" },
  { code: "CM", label: "Centímetro" },
  { code: "MM", label: "Milímetro" },
  { code: "M2", label: "Metro cuadrado" },
  { code: "M3", label: "Metro cúbico" },
  { code: "FT", label: "Pie" },
  { code: "IN", label: "Pulgada" },
  { code: "LT", label: "Litro" },
  { code: "ML", label: "Mililitro" },
  { code: "GL", label: "Galón" },
  { code: "PR", label: "Par" },
  { code: "JG", label: "Juego" },
  { code: "KT", label: "Kit" },
  { code: "CJ", label: "Caja" },
  { code: "RL", label: "Rollo" },
  { code: "HJ", label: "Hoja" },
];

export const UOM_CODES = new Set(UOM_OPTIONS.map((u) => u.code));

export function uomLabel(code: string): string {
  const u = UOM_OPTIONS.find((o) => o.code === code);
  return u ? `${u.code} — ${u.label}` : code;
}
